tests/conftest.py

```diff
@@ -119,11 +118,31 @@ def generate_tests(self, smoke_test_data, random_test_data=[]):
         _: list
             Tests.
         """
-        smoke_tests = [
-            pytest.param(*data.values(), marks=smoke) for data in smoke_test_data
-        ]
-        random_tests = [pytest.param(*data, marks=random) for data in random_test_data]
-        return smoke_tests + random_tests
+        tests = []
+        if smoke_test_data:
+            smoke_tests = [
+                pytest.param(*data.values(), marks=pytest.mark.smoke)
+                for data in smoke_test_data
+            ]
+            tests += smoke_tests
+        if random_test_data:
+            random_tests = [
+                pytest.param(*data, marks=pytest.mark.random)
+                for data in random_test_data
+            ]
+            tests += random_tests
+        return tests
+
+    def log_exp_composition(self, space, num_samples=10, max_n=10, num_n=5, **kwargs):
```

Comment:
You pass the class `space` as argument of the function, as opposed to an instantiated object. Is it because the error message is easier to read this way? Otherwise, why passing the class and not the instantiated object?

Reply:
if we pass space, then we can't `n` parametrize directly. so when error happens we don't know at which `n` it is failing

Reply:
Yes, thanks for re-explaining this to.